src/timer.c

```diff
@@ -335,16 +397,50 @@ void timer_close(void)
 }
 
 /**
- * @brief Return total ticks since timer was initialized
+ * @brief Return total ticks since timer was initialized, as a 64-bit counter.
  *
  * @return Then number of ticks since the timer was initialized
+ *
  */
 long long timer_ticks(void)
 {
-	disable_interrupts();
-	timer_callback();					// force processing the timers
-	enable_interrupts();
-	return total_ticks;
+	uint32_t low, high;
+
+	/* Read status register */
+    uint32_t sr;
+    asm volatile("mfc0 %0,$12":"=r"(sr));
+
+	/* Check whether interrupts are enabled or not. We need a different strategy
+	 * to account for race conditions. */
+	if (sr & 1) {
+		/* Read the hardware counter twice, and fetch the high part counter
+		 * in between. In the unlikely case that the counter overflows exactly
+		 * during the sequence, it means that there's a race condition and
+		 * we can't really know whether high and low are coherent -- but in
+		 * that case, we just repeat the sequence again to avoid the ambiguity. */
+		uint32_t pre;
+		do {
+		    pre = TICKS_READ();
+		    MEMORY_BARRIER();
+		    high = ticks64_high;
+		    MEMORY_BARRIER();
+		    low = TICKS_READ();
+		} while ((int32_t)pre < 0 && (int32_t)low >= 0);
+
+	} else {
+		/* Interrupts are currently disabled. If they've been disabled for more
+		 * than 2**32 ticks, it's game over because we can't know how many times
+		 * the counter has overflown.
+		 * So assuming they were disabled for not too long, check whether there
+		 * was a counter overflow between now and the last time an
+		 * interrupt triggered. If there was, increment high. */
+	    low = TICKS_READ();
+	    high = ticks64_high;
+		if (last_interrupt_tick > low)
```

Comment:
No actually it doesn't require such a long time. Consider this sequence:

 * FFFFFE00: VI interrupt triggers (last interrupt)
 * FFFFFE88: user disables interrupts
 * 00000010: user calles `timer_ticks()`
 
With the code as in the PR, `timer_ticks()` will return `(high+1) << 32 + 0x10` which is indeed correct because an interrupt occurred. The code realizes that an interrupt occurred because it knows that the timer was recently FFFFFE00, so a timer of 10 indicates an overflow. On the contrary:

 * 00000100: VI interrupt triggers (last interrupt)
 * 00000188: user disables interrupts
 * 000001AA: user calles `timer_ticks()`

In this case, `high` would not be incremented because the timer was recently 0x100, so 0x1AA has not overflown yet.

Reply:
Ah, I thought we were updating `last_interrupt_tick` only if it is a timer interrupt. Why not only update it for a timer interrupt? Then you're still right there may be many timers registered that may trigger at any time.

Reply:
`last_interrupt_tick` is just a recent timer snapshot, so it doesn't really matter how often we update it. 

Thinking more of it, what we really want is to know when interrupts were disabled. I've changed `last_interrupt_tick` into `interrupts_disabled_tick`, modified only by `disable_interrupts()`. This fails if the user disables interrupts without going through `disable_interrupts()` but I think that's already not fully supported anyway, so we might ignore it.

I hope it's more clear now. It also allows to remove the `last_interrupt_tick` modification from the test when the timer is jumping around, as we just now record the correct value when disabling the interrupts anyway.